### DISCUSSION TITLE
Unterstützung von utf8 beim PDF Etiketten:  mPDF anstatt FPDF

### DISCUSSION THREAD

**Discussion by dmitry-vs2019** [Larian77/Openbiblio#37] [General]
Hallo Openbiblio-Team,
Ich habe bemerkt, dass Human Readable Text bei PDF-Barcode-Etiketten nicht mitgedruckt wird, wenn der Text einige Symbolen aus zum Beispiel  Hebreisch oder Russisch enthält. 
Die FPDF Bibliothek, die in Openbiblio benutzt, unterstützt uft8 in PDF nicht komplett, deswegen es ist sinnvoll, die Bibliothek mPDF (https://mpdf.github.io/) zu benutzen.
Habt ihr vor, dass zu machen?

**Comment by Larian77** [comment]
bisher haben wir das noch nicht beachtet, ist aber gut vorstellbar dass wir das machen - muss ich mir aber etwas genauer anschauen erstmal :-)

**Comment by dmitry-vs2019** [comment]
Hallo Openbiblio-Team,
in der Zwischenzeit habe ich versucht, die Openbiblio mit der mPDF zu verbinden, das hat in der ersten Approximation funktioniert. Das System ist Windows 10 Home, XAMPP mit PHP-Version: 8.2.12, Openbiblio 0.8.0 Um das zu ermöglichen, brauchen wir:
1.	Die mPDF-Bibliothek installieren z.B. in htdocs/pdf-Verzeichnis mit Hilfe von composer: 
https://github.com/mpdf/mpdf
https://mpdf.github.io/installation-setup/installation-v7-x.html 
und Bacode-Support auch installieren.

2.	Eine System-Umgebungsvariable MPDF_ROOT erstellen, die den Pfad zum mPDF enthält.

3.	Die neuen überarbeiteten Dateien mpdf_labels.php und mpdf_list.php. in layout/default-Verzeichnis kopieren 

4.	In shared-Verzeichnis: die existierende Datei layout.php umbenennen in z.B. old_layout.php zum Backup und die neue layout.php hierhin kopieren. Die neue überarbeitete layout.php erkennt die installerte mPDF und passt die Dateinamen und Klassennamen für die Etiketten an. Falls die mPDF-Bibliothek nicht installiert wurde, funktioniert die layout.php wie früh.
Als Ergebnis haben wir:
1.	Eine PDF-Etiketten-Datei mit dem Layout 3x10 für A4
2.	Eine PDF-Liste-Datei, die auch auf A4 Seiten geteilt wird
Alle utf8-Symbolen sind da und die Barcodes sind mit dem Barcodescanner Symbol LS2208 (Hergestellt 2004, Mexico) lesbar!
Es wurde kaum speziellen Einstellungen, keine speziellen Formatierungen für mPDF-Bibliothek benötigt! Wir haben sehr schnell praktische Ergebnisse bekommen.
Ich habe einen Vorschlag: die FPDF-Bibliothek, die bei existierender Openbiblio-Version benutzt wird, durch die mPDF-Bibliothek zu ersetzen. Vorteile: komplette Unterstützung von utf8/UNICODE, leichte intuitive Interface, die PDF-Erstellung direkt von HTML-Seiten, lesbare Barcode 😊
Ich möchte bei eurem Projekt mitwirken, wenn das geht. Ich bin erfahrener freiberuflicher C/C++-Programmierer. Mit PHP- und Web-Programmierung habe ich weniger Erfahrung.

Mit freundlichen Grüßen.
Dmitry

**Comment by dmitry-vs2019** [comment]
Anbei sind die 3 obengenannte php-Dateien: 
[mpdf_layout.zip](https://github.com/Larian77/Openbiblio/files/15370202/mpdf_layout.zip)

**Comment by LeAchim59** [comment]
Bei uns läuft openBiblio auf einem Webserver gehostet von Strato. mpdf steht da nicht zur Verfügung. Die Bibliothek kann ich vielleicht noch installieren, aber System-Umgebungsvariablen kann ich da (soweit ich weiß) nicht setzen.

**Reply by dmitry-vs2019** [reply]
OK, die System-Umgebungsvariable ist nicht notwendig. Ich wollte mpdf für mehrere Projekte benutzen. Man kann sogar die mdpf-Bibliothek bei der Installation von Openbiblio ins Unterverzeichnis Openbiblio/mpdf mit kopieren oder separat da installieren. Nur entsprechende Pfade zur mpdf bei mpdf_labels.php und mpdf_list.php  sollten angepasst werden. Man kann parallel für Etiketten  die beide Bibliotheken  FPDF und mPDF verwenden, aber dafür soll man eine Etiketten-Menü erweitern. Ich kann das innerhalb die nächste Woche vorbereiten.

**Reply by 375michael40veit** [reply]
Wenn mpdf mit in den Ordner openbiblio integriert wird, dann sollte die Bibliothek unter openbiblio/lib/ gespeichert werden. Im Zuge der Integration von phpMailier, welcher unter openbiblio/lib/ gespeichert wird, hätten wir dann wieder eine sinnvolle Ordnerstruktur.

**Comment by Larian77** [comment]
wenn dann muss wie schon erwähnt mpdf auf alle Fälle mit ins openbiblio-Paket, damit keine weitere Bibliothek extra installiert werden muss, wenn diese nicht überall bereit gestellt wird, sollten wir auf alle Fälle angehen :-)

**Comment by superudo** [comment]
Es ist durchaus möglich, mpdf komplett ins Paket zu integrieren - in ähnlicher Weise, wie es mit anderen Bibliotheken bereits erfolgt ist. Im Branch https://github.com/superudo/Openbiblio/tree/enhancement/mpdf meines Forks habe ich den Vorgang bereits einmal durchgeführt. Eine Testseite /test_mpdf.php zeigt die Funktion. Das gesamte Paket vergrößert das Projekt um ca. 100MB, ist dann aber ohne weitere Installationen betriebsbereit. Näheres in der [README-MPDF.md](https://github.com/superudo/Openbiblio/blob/enhancement/mpdf/README-MPDF.md) im Branch.

---

It is definitely possible to integrate mpdf completely into the package — in a similar manner to how other libraries have already been integrated. In the branch https://github.com/superudo/Openbiblio/tree/enhancement/mpdf of my fork, I have already performed this process once. A test page /test_mpdf.php demonstrates the functionality. The entire package increases the project by approximately 100MB, but then it is ready to operate without any further installations. Further details can be found in the [README-MPDF.md](https://github.com/superudo/Openbiblio/blob/enhancement/mpdf/README-MPDF.md) in the branch.
